internal/builtinfunctions/functions.go

```diff
@@ -0,0 +1,428 @@
+			),
+			nil,
+		),
+		func(a float64) int64 {
+			return int64(a)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	return funcSchema
+}
+
+func getIntToStringFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"intToString",
+		[]schema.Type{schema.NewIntSchema(nil, nil, nil)},
+		schema.NewStringSchema(nil, nil, regexp.MustCompile(`^(?:0|[1-9]\d*)$`)),
+		false,
+		schema.NewDisplayValue(
+			schema.PointerTo("intToString"),
+			schema.PointerTo(
+				"Converts an integer to a string whose characters represent that integer in base-10.\n"+
+					"For example, an input of `55` will output \"55\"",
+			),
+			nil,
+		),
+		func(a int64) string {
+			return strconv.FormatInt(a, 10)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	return funcSchema
+}
+
+func getFloatToStringFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"floatToString",
+		[]schema.Type{schema.NewFloatSchema(nil, nil, nil)},
+		schema.NewStringSchema(nil, nil, regexp.MustCompile(`^\d+\.\d*$`)),
+		false,
+		schema.NewDisplayValue(
+			schema.PointerTo("floatToString"),
+			schema.PointerTo(
+				"Converts a floating point number to a string whose characters"+
+					"represent that number in base-10 as as simple decimal.\n"+
+					"For example, an input of `5000.5` will output \"5000.5\"",
+			),
+			nil,
+		),
+		func(a float64) string {
+			return strconv.FormatFloat(a, 'f', -1, 64)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	return funcSchema
+}
+
+// TODO: Webb can add a function called floatToStringAdvanced that allows other float formats
+
+func getBooleanToStringFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"boolToString",
+		[]schema.Type{schema.NewBoolSchema()},
+		schema.NewStringSchema(nil, nil, regexp.MustCompile(`^true|false$`)),
+		false,
+		schema.NewDisplayValue(
+			schema.PointerTo("boolToString"),
+			schema.PointerTo(
+				"Represents `true` as \"true\", and `false` as \"false\".",
+			),
+			nil,
+		),
+		strconv.FormatBool, // Wrap go standard lib function.
+	)
+	if err != nil {
+		panic(err)
+	}
+	return funcSchema
+}
+
+func getStringToIntFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"stringToInt",
+		[]schema.Type{schema.NewStringSchema(nil, nil, regexp.MustCompile(`^-?(?:0|[1-9]\d*)$`))},
+		schema.NewIntSchema(nil, nil, nil),
+		true,
+		schema.NewDisplayValue(
+			schema.PointerTo("stringToInt"),
+			schema.PointerTo(
+				"Interprets the string as a base-10 integer. Can fail if the input is not a valid integer.",
+			),
+			nil,
+		),
+		func(s string) (int64, error) {
+			return strconv.ParseInt(s, 10, 0)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	return funcSchema
+}
+
+func getStringToFloatFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"stringToFloat",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)},
```

Comment:
As [I said above](https://github.com/arcalot/arcaflow-engine/pull/150#discussion_r1492716633), validating the output/input _values_ is probably not valuable once a workflow is put into production (that is, there is a cost to it which may not be worth paying if the check never fails, and it may be more practical to detect a failure further downstream).

What is of interest is being able to detect incompatibilities between the output and input before the workflow starts running.  But, in order to do that, the respective schemas need to be accurate, particularly the input one.  So, if we're going to perform that comparison, then we should have an appropriate restriction here.

FWIW, you mention learning by looking at the schema...is there a way to access the schema other than by looking at the code here?  If not, I don't think that argument holds much weight, because I wouldn't expect most users or plugin developers to dive into the source in order to find the function requirements.

Reply:
It is possible to export plugin schemas to a YAML format. Workflows define their own schemas in a YAML format.

Reply:
Some checks exist. I think it's out of the scope of this discussion.